### PR TITLE
EOS Support

### PR DESCRIPTION
**Pull request by keepkeyjon**
Consider this a **very** rough cut at adding EOS support. I've still got **lots** of cleaning up to do, and haven't tested it at all. That said, I'd appreciate early feedback, particularly on my use of `layoutDialogSwipe()`... it's a bit less easy to use than our `confirm()`.

### CODE REVIEW

File: firmware/eosio_token.c [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        _("Cancel"),
+        _("Confirm"),
+        _("Transfer"),
+        _("Do you want to send"),
+        asset,
+        _("from"),
+        sender,
+        _("to"),
+        receiver);
+    if (!protectButton(ButtonRequestType_ButtonRequest_ConfirmOutput, false)) {
```
Comment (keepkeyjon) [inline firmware/eosio_token.c]: In our port, I called this button request `ConfirmAction`. What would you prefer here? It seems unfriendly to clients to call everything `ConfirmOutput`...

File: firmware/eos.c [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+#include "hasher.h"
+#include "memzero.h"
+#include "secp256k1.h"
+
+#include "messages-eos.pb.h"
+
+#include <stdio.h>
+#include <time.h>
+
+Hasher hasher_preimage;
```
Comment (keepkeyjon) [inline firmware/eos.c]: In ours I marked these CONFIDENTIAL out of an abundance of caution. When I do that here, the section ends up too big.

File: firmware/Makefile [hunk 3]
```diff
@@ -128,3 +132,6 @@ bl_data.h: bl_data.py ../bootloader/bootloader.bin
 clean::
 	rm -f bl_data.h
 	find -maxdepth 1 -name "*.mako" | sed 's/.mako$$//' | xargs rm -f
+
```
Comment (keepkeyjon) [inline firmware/Makefile]: Note to self: delete stray whitespace.

File: firmware/eos.c [hunk 4]
```diff
@@ -0,0 +1,476 @@
+#include "eos.h"
```
Comment (keepkeyjon) [inline firmware/eos.c]: What license header should I put on the new files?

File: firmware/eosio_system.c [hunk 5]
```diff
@@ -0,0 +1,932 @@
...
+    eos_hashUInt(&hasher_preimage, size);
+
+    hasher_Update(&hasher_preimage, (const uint8_t*)&action->account, 8);
+    hasher_Update(&hasher_preimage, (const uint8_t*)&action->code, 8);
+    hasher_Update(&hasher_preimage, (const uint8_t*)&action->type, 8);
+
+    return true;
+}
+
+bool eos_compileActionNewAccount(const EosActionCommon *common,
```
Comment (keepkeyjon) [inline firmware/eosio_system.c]: Note to self: add back the SLIP48 stuff.

File: firmware/eos.c [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+        case EOS_BuyRam:
+        case EOS_BuyRamBytes:
+        case EOS_SellRam:
+        case EOS_VoteProducer:
+        case EOS_UpdateAuth:
+        case EOS_DeleteAuth:
+        case EOS_LinkAuth:
+        case EOS_UnlinkAuth:
+            return true;
+        }
```
Comment (keepkeyjon) [inline firmware/eos.c]: Note to self: add `__builtin_unreachable()` to force enum additions to also be added to the switch.

File: firmware/eos.c [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+bool eos_getPublicKey(const HDNode *n, char *pubkey, size_t len) {
+    return eos_publicKeyToWif(n->public_key, pubkey, len);
+}
+
+bool eos_publicKeyToWif(const uint8_t *public_key, char *pubkey, size_t len) {
+    const char *prefix = "EOS";
+    const size_t prefix_len = strlen(prefix);
+    strlcpy(pubkey, prefix, len);
+
+    if (!base58_encode_check(public_key, 33, HASHER_SHA2, // FIXME: HASHER_RIPEMD,
```
Comment (keepkeyjon) [inline firmware/eos.c]: Reminds me I need to make a patch for trezor-crypto.

**Comment by keepkeyjon** [comment]
Depends on: https://github.com/trezor/trezor-common/pull/231

**Comment by prusnak** [comment]
We won't be adding EOS to trezor-mcu in the end. We will add EOS to Trezor One by porting Trezor Core to Trezor One.

**Comment by keepkeyjon** [comment]
ok

**Pull request closed without merging**